dcrec/secp256k1/signature.go

```diff
@@ -509,31 +506,40 @@ func RecoverCompact(signature, hash []byte) (*PublicKey, bool, error) {
 func signRFC6979(privateKey *PrivateKey, hash []byte) *Signature {
 	curve := S256()
 	N := order
+	var kBytes [32]byte
+	defer zeroArray32(&kBytes)
 	privKeyBytes := privateKey.key.Bytes()
 	bigPrivKey := new(big.Int).SetBytes(privKeyBytes[:])
 	defer zeroArray32(&privKeyBytes)
 	defer bigPrivKey.SetUint64(0)
 	for iteration := uint32(0); ; iteration++ {
 		k := NonceRFC6979(privKeyBytes[:], hash, nil, nil, iteration)
-		inv := new(big.Int).ModInverse(k, N)
-		r, _ := curve.ScalarBaseMult(k.Bytes())
+		k.PutBytes(&kBytes)
+		bigK := new(big.Int).SetBytes(kBytes[:])
+		r, _ := curve.ScalarBaseMult(kBytes[:])
 		r.Mod(r, N)
 		if r.Sign() == 0 {
+			bigK.SetUint64(0)
 			continue
 		}
 
+		kInv := new(big.Int).ModInverse(bigK, N)
 		e := hashToInt(hash)
 		s := new(big.Int).Mul(bigPrivKey, r)
 		s.Add(s, e)
-		s.Mul(s, inv)
+		s.Mul(s, kInv)
 		s.Mod(s, N)
 
 		if s.Cmp(halforder) == 1 {
 			s.Sub(N, s)
 		}
 		if s.Sign() == 0 {
+			bigK.SetUint64(0)
+			kInv.SetUint64(0)
 			continue
 		}
+		bigK.SetUint64(0)
+		kInv.SetUint64(0)
```

Comment:
are these being zerod to clear the key out of memory?  Because this method is not sufficient for this. You need to call Bits() and zero that.

Reply:
Nice catch.  I've updated this to introduce a new function `zeroBigInt` which zeros the individual words returned by `Bits` and updated all of these call sites accordingly.